File: C++/Lanchonete/Aplicativo/Pedido.cpp
#include <iostream>
#include <string>
#include <vector>
#include "Pedido.h"

using namespace std;

Pedido::Pedido(int &codigo, Cliente *cliente, string &data) {
    this->codigo = codigo;
    this->cliente = cliente;
    this->data = data;

    produtos = new vector<Produto *>();
    quantidadeProdutos = new vector<int>();
}

Pedido::Pedido(int &codigo, Cliente *cliente) {
    string date = __DATE__;
    string time = __TIME__;
    string dataCompleta = date + " " + time;

    this->codigo = codigo;
    this->cliente = cliente;
    this->data = dataCompleta;

    produtos = new vector<Produto *>();
    quantidadeProdutos = new vector<int>();
}

Pedido::~Pedido(){
    delete produtos;
    delete quantidadeProdutos;
}

void Pedido::imprimirDetalhesPedido() {
    cout << "\nCódigo=" << codigo << ", Cliente=" << cliente->getNome() << ", Data=" << data << "\n" << endl;

    for (int i = 0; i < produtos->size(); i++) {
        cout << "Produto=" + produtos->at(i)->getNome() <<
             ", Quantidade=" << quantidadeProdutos->at(i) <<
             ", Valor Unitario=" << produtos->at(i)->getPreco() <<
             ", Valor Total=" << (produtos->at(i)->getPreco() * quantidadeProdutos->at(i)) << endl;
    }

    cout << "\nValor total do pedido: R$" << this->valorTotal() << endl;
}

void Pedido::addProduto(Produto *p, int quantidade) const {
    this->produtos->push_back(p);
    this->quantidadeProdutos->push_back(quantidade);
}

double Pedido::valorTotal() const {
    double valor = 0;

    for (int i = 0; i < produtos->size(); i++) {
        valor += (produtos->at(i)->getPreco() * quantidadeProdutos->at(i));
    }

    return valor;
}

int Pedido::quantidadeDeProdutos() const {
    int qtd = 0;

    for (int x: *quantidadeProdutos) {
        qtd += x;
    }

    return qtd;
}

int Pedido::getCodigo() const {
    return codigo;
}

Cliente *Pedido::getCliente() {
    return cliente;
}

string Pedido::getData() {
    return data;
}

ostream &operator<<(ostream &stream, const Pedido &p) {
    stream << "Pedido{" <<
         "Codigo=" << p.codigo <<
         ", Cliente=" << p.cliente->getNome() <<
         ", Data='" << p.data + '\'' <<
         ", Quantidade de Produtos=" << p.quantidadeDeProdutos() <<
         ", Valor Total=" << p.valorTotal() <<
         '}';

    return stream;
}

File: Java/Lanchonete/src/Aplicativo/Pedido.java
package Aplicativo;

import Cadastro.Cliente;

import java.time.LocalDateTime;
import java.time.format.DateTimeFormatter;
import java.util.ArrayList;

public class Pedido {
    private final int codigo;
    private final Cliente cliente;
    private final String data;
    public ArrayList<Produto> produtos;
    public ArrayList<Integer> quantidadeProdutos;

    public Pedido(int codigo, Cliente cliente, String data){
        this.codigo = codigo;
        this.cliente = cliente;
        this.data = data;

        produtos = new ArrayList<>();
        quantidadeProdutos = new ArrayList<>();
    }
    public Pedido(int codigo, Cliente cliente){

        //Pega a data atual no formato dia/mes/ano horas:minutos:segundos
        this(codigo, cliente, DateTimeFormatter.ofPattern("dd/MM/yyyy HH:mm:ss").format(LocalDateTime.now()));

    }

    @Override
    public String toString() {
        return "Pedido{" +
                "Codigo=" + codigo +
                ", Cliente=" + cliente.getNome() +
                ", Data='" + data + '\'' +
                ", Quantidade de Produtos=" + quantidadeDeProdutos() +
                ", Valor Total=" + valorTotal() +
                '}';
    }

    public void imprimirDetalhesPedido(){
        System.out.println("\nCódigo=" + codigo + ", Cliente=" + cliente.getNome() + ", Data=" + data + "\n");

        for(int i=0; i<produtos.size(); i++){
            System.out.println("Produto=" + produtos.get(i).getNome() +
                                ", Quantidade=" + quantidadeProdutos.get(i) +
                                ", Valor Unitario=" + produtos.get(i).getPreco() +
                                ", Valor Total=" + (produtos.get(i).getPreco() * quantidadeProdutos.get(i)) );
        }

        System.out.println("\nValor total do pedido: R$" + this.valorTotal());
    }

    //Adiciona um produto ao pedido
    void addProduto(Produto p, int quantidade){
        this.produtos.add(p);
        this.quantidadeProdutos.add(quantidade);
    }

    //Retorna valor total do pedido
    public double valorTotal(){
        double valor = 0;

        for(int i=0; i<produtos.size(); i++){
            valor += (produtos.get(i).getPreco() * quantidadeProdutos.get(i));
        }

        return valor;
    }

    //Retorna a quantidade de produtos totais do pedido
    public int quantidadeDeProdutos(){
        int qtd = 0;

        for(int x : quantidadeProdutos){
            qtd += x;
        }

        return qtd;
    }

    public int getCodigo() {
        return codigo;
    }

    public Cliente getCliente() {
        return cliente;
    }

    public String getData() {
        return data;
    }
}
